gensim/models/poincare.py

```diff
@@ -0,0 +1,851 @@
+            for item in hypernym_pair:
+                if item in vocab:
+                    vocab[item].count += 1
+                else:
+                    vocab[item] = Vocab(count=1, index=len(index2word))
+                    index2word.append(item)
+            node_1, node_2 = hypernym_pair
+            node_1_index, node_2_index = vocab[node_1].index, vocab[node_2].index
+            node_relations[node_1_index].add(node_2_index)
+            relation = (node_1_index, node_2_index)
+            all_relations.append(relation)
+        logger.info("Loaded %d relations from train data, %d unique terms", len(all_relations), len(vocab))
+        self.kv.vocab = vocab
+        self.kv.index2word = index2word
+        self.indices_set = set((range(len(index2word))))  # Set of all node indices
+        self.indices_array = np.array(range(len(index2word)))  # Numpy array of all node indices
+        counts = np.array([self.kv.vocab[index2word[i]].count for i in range(len(index2word))], dtype=np.float64)
+        self._node_probabilities = counts / counts.sum()
+        self._node_probabilities_cumsum = np.cumsum(self._node_probabilities)
+        self.all_relations = all_relations
+        self.node_relations = node_relations
+        self._negatives_buffer = NegativesBuffer([])  # Buffer to store negative samples, to reduce calls to sampling method
+        self._negatives_buffer_size = 2000
+
+    def _init_embeddings(self):
+        """Randomly initialize vectors for the items in the vocab."""
+        shape = (len(self.kv.index2word), self.size)
+        self.kv.syn0 = self._np_random.uniform(self.init_range[0], self.init_range[1], shape)
+
+    def _get_candidate_negatives(self):
+        """Returns candidate negatives of size `self.negative` from the negative examples buffer.
+
+        Returns
+        --------
+        numpy.array
+            Array of shape (`self.negative`,) containing indices of negative nodes.
+
+        """
+
+        if self._negatives_buffer.num_items() < self.negative:
+            # Note: np.random.choice much slower than random.sample for large populations, possible bottleneck
+            uniform_numbers = self._np_random.random_sample(self._negatives_buffer_size)
+            cumsum_table_indices = np.searchsorted(self._node_probabilities_cumsum, uniform_numbers)
+            self._negatives_buffer = NegativesBuffer(cumsum_table_indices)
+        return self._negatives_buffer.get_items(self.negative)
+
+    @staticmethod
+    def _has_duplicates(array):
+        """Returns whether or not the input array has any duplicates.
+
+        Parameters
+        ----------
+        array : iterable of hashables
+            Input array to checked, should contain hashable items.
+
+        Returns
+        -------
+        bool
+            Whether the input array contains any duplicates.
+
+        """
+        seen = set()
```

Comment:
`return len(array) != len(set(array))` simpler. Probably not even worth adding a method.

Reply:
Done